### DISCUSSION TITLE
Named Params for Generic Action/Func: e.g. "event Action<string name, int age> EventName"

### DISCUSSION THREAD

**Discussion by najak3d** [dotnet/csharplang#5835] [Language Ideas] [answered]
Is there a reason this notation is not a good idea:

      public event Action<string name, int age> BirthdayOccurred;

Then the Intellisense and default function params auto-created for you all make sense instead of showing "args1, args2, ..."

I realize this can sorta be accomplished now using Tuple-notation, like this:

      public event Action<(string name, int age)> BirthdayOccurred;

But then on the receiving side, the method is a bit weird, accepting a tuple.  We'd prefer to avoid using Tuples to achieve this convenient named parameters effect.

**Comment by JustNrik** [comment]
Because events return delegates and delegates don't force anyone to use specific parameter names, you can't force anyone to name an event parameter as you want it unless you use an analyzer, with tuple it's the same, you can use named tuples, but people who subscribe to the event can name their tuples in any way they like or don't name them at all.

**Reply by najak3d** [reply]
> Because events return delegates and delegates don't force anyone to use specific parameter names, you can't force anyone to name an event parameter as you want it unless you use an analyzer, with tuple it's the same, you can use named tuples, but people who subscribe to the event can name their tuples in any way they like or don't name them at all.

@JustNrik  - my proposal isn't for "force naming on the recipient" but would simply be there for "comprehension" so that a person knows the meaning of each argument (without forcing the creation of a strong-typed delegate to accomplish this).

Action<string, int, bool>  -- what are these arguments?

Action<string name, int age, bool isMarried>   <===  now it's clear!

Since the latter is not possible, it leaves Action/Func<T> methods deficient on clarity, which is unfortunate.


Similar to how it's done with Tuples, the parameter names would only be used to:
1. Populate Intellisense
2. When the IDE auto-generates the handler method for you, instead of having "(string arg1, int arg2, bool arg3)" it'll generate meaningful parameter names.

**Reply by CyrusNajmabadi** [reply]
> Action<string, int, bool> -- what are these arguments?

It's been a core point since we added Action/Func that you should only use them when their types are apparent from context, without names. If it's not apparent, create a custom delegate type.

**Reply by theunrepentantgeek** [reply]
If your motivating example is this:

``` csharp
Action<string name, int age, bool isMarried> 
```

I'd argue that you should be writing instead:

``` csharp
Action<Person> 
```

Failing to recognise domain types is a fertile source of bugs that no amount of fussing with the language will remedy.

**Reply by FaustVX** [reply]
As a workaround, you can use tuple, it's not ideal but it works:
```csharp
public event Action<(string name, int age)> BirthdayOccurred;
```

**Reply by najak3d** [reply]
@theunrepentantgeek :   How does this proposal differ from Tuple parameter naming? 

I agree, my example was lame (i.e Action\<Person> is the best solution there).

How about this one:
  public event <Person person, string oldName, string newName> NameChanged;

without parameter names, this is simply:
  public event <Person, string, string> NameChanged;

The first entry is superior to creating a strong-named delegate, for one-use only.   Without parameter names, it puts you in a conundrum -- to create a new delegate Type for one-use?  Or use the Action<Person, string, string> and put the parameter meaning into the \<summary> comment?   (I do the latter, because I don't want to clutter my namespaces with one-usage delegates.)

I see zero downside to this proposal; only upside.

**Reply by najak3d** [reply]
> As a workaround, you can use tuple, it's not ideal but it works:
> 
> ```cs
> public event Action<(string name, int age)> BirthdayOccurred;
> ```

I realized this workaround, and may start using it.     But using that method begs the question of why introduce a Tuple, where it shouldn't be required, if C# syntax simply granted the same provisions to anonymous delegates as it does for anonymous records.  Thus, why I am here, barking up this tree -- hoping a champion hears my plea. :)

**Reply by HaloFour** [reply]
There are thousands of language proposals on this repo, most offering a little bit of upside.  Making any change to the language is in of itself already a fairly big downside, so the upside has to be demonstatively greater than that.

**Reply by najak3d** [reply]
@HaloFour  - my experience has been that anonymous delegates are used rampantly -- and of course ALL have this inferiority of ambiguous parameter names.   This isn't a "specialized/niche" request - it's very mainstream.   And the request also seems rather small/easy in nature -- as the parameter names wouldn't even change functionality, but simply serves as built-in documentation - that shows up in Intellisense and is used for auto-generation of the handler methods, etc.

This one seems to me as "low-hanging fruit" (significant gain, for very little cost)

**Reply by HaloFour** [reply]
This proposal does zero to change that because the delegates parameters are not reflected in the lambda.  As such there is no gain here.

**Reply by CyrusNajmabadi** [reply]
> This one seems to me as "low-hanging fruit" (significant gain, for very little cost)

I disagree with this. This would be high cost for little gain. We'd have to define the way to even encode this in metadata. We'd have to come up with rules about what is or isn't allowed (like we did with tuples). Etc. etc.  For example can you assign a Func with one set of names to another with a different set of names?  Is that a warning?

And ask this just to get names that you can already get with something that's an easy one liner. 

The comparison to tuples is not analogous here as tuples provide a wealth of other stock functionality beyond field naming. Prior to tuples there was no way to get this yourself without writing dozens of lines of code each time.

**Reply by foxesknow** [reply]
> > Action<string, int, bool> -- what are these arguments?
> 
> It's been a core point since we added Action/Func that you should only use them when their types are apparent from context, without names. If it's not apparent, create a custom delegate type.

You could make the same point about tuples and yet they support naming.

**Reply by CyrusNajmabadi** [reply]
That's not true though.  When we added tuples it was not a core point of them that they should be used when their types (absent names) make them apparent from context.   Indeed, when tuples were created there was no alternative that would give you taht with all the rest of the benefits of tuples.

**Reply by foxesknow** [reply]
> There are thousands of language proposals on this repo, most offering a little bit of upside. Making any change to the language is in of itself already a fairly big downside, so the upside has to be demonstatively greater than that.

And yet someone someone signed off on expression methods when there was already a perfectly good way of writing methods. What's was the presumed upside here?

**Reply by CyrusNajmabadi** [reply]
> And yet someone someone signed off on expression methods when there was already a perfectly good way of writing methods. What's was the presumed upside here?

The upside was that expression bodied methods were nicer to write for a very common case.

**Reply by foxesknow** [reply]
> > And yet someone someone signed off on expression methods when there was already a perfectly good way of writing methods. What's was the presumed upside here?
> 
> The upside was that expression bodied methods were nicer to write for a very common case.

That's debatable. 

Conversely, the downside is you've now got two ways of doing the same thing, where the new way need to be transformed once you can't express intent as a single expression. It also causes people to write overly cryptic code in order to cram everything into a single expression.

But hey, you don't have to write `return` any more...

**Reply by CyrusNajmabadi** [reply]
> That's debatable.

All language changes are. :)

**Reply by CyrusNajmabadi** [reply]
>  also causes people to write overly cryptic code in order to cram everything into a single expression.

Nothing about the prior way prevents that. People still write overly cryptic code using the old style.

**Comment by iam3yal** [comment, marked as answer]
See #894, #1883, #269 and #5706.

**Reply by najak3d** [reply]
#1883 - IMO, of all those threads, describes this the best, and most comprehensively.

**Comment by najak3d** [comment]
**#1883 - is 3.5 years old, and dormant.**

I think it's worth reviving, and championing.   What is the process for trying to revive an old/dormant idea like this?

**Reply by HaloFour** [reply]
> What is the process for trying to revive an old/dormant idea like this?

It's up to a member of the language team to determine that there is value in the proposal.  Without a champion a proposal doesn't go anywhere.

**Reply by najak3d** [reply]
How does one summon one of these Champions?   I'm not sure who on this site is a Champion, vs. not.

What is your opinion of this proposal?  #1883    I believe it is highly Champion-worthy.  Seems to be low-hanging fruit (high return for not much work, compared to other proposals).

**Reply by CyrusNajmabadi** [reply]
> How does one summon one of these Champions

They read the proposals and decide which are worth doing. I'm one of them for example. Imo this isn't worth doing as I can there is already a suitable answer here with normal delegates. Action/Func exist precisely for the case of: these params don't need names as they are obvious in context. 

If they are not obvious, I would recommend not using Action/Func.

**Comment by najak3d** [comment]
@HaloFour & @CyrusNajmabadi :  - example of gain here:

```c#
// =========== Proposed Solution ============== 
public event Action<Person employee, Person oldManager, Person newManager> ManagerChanged;


// =========== Current Solution =============
namespace SomeOtherNamespace   /// and maybe in another DLL
{
  public delegate void PersonRelationshipChangedDelegate(Person employee, Person oldManager, Person newManager);
}
// === in another file, maybe another DLL
namespace MyNameSpace
{
  public event SomeOtherNamespace.PersonRelationshipChangedDelegate ManagerChanged;
}
```

**Now look at the Intellisense result - Present Solution**
![image](https://user-images.githubusercontent.com/10914515/155986465-55b3a359-ca56-4001-9d7a-00d18d0881cb.png)

**Proposed Solution Intellisense: (facilitating via Tuple syntax for now)**
![image](https://user-images.githubusercontent.com/10914515/155986531-73af0aa1-6fb0-4f7e-b0d1-b44a902f96b3.png)


As you can see, not only is the proposed solution easier for the coder to write, but also is easier for the consumer of this event to see the arguments of the event/delegate.   The proposed solution make it clear what the delegate is going to do from the Name + Parameter Names.

**Reply by HaloFour** [reply]
We're well aware of what the code would look like today.  As for what Intellisense presents, if you'd rather see a more complete signature that would be a tooling request and not something that needs to be solved by the language.  When reading code you would not see that information, nor would those parameter names be preserved anywhere.

As an aside, the recommendation is to use `EventHandler<T>` delegates with events, not one of anonymous delegate types.

**Reply by HaloFour** [reply]
A named tuple as the generic argument of the delegate also seems to achieve what you want and has the added benefit of allowing the subscriber of the event dot off of those names:

```cs
public event Action<(Person employee, Person oldManager, Person newManager)> ManagerChanged;


obj.ManagerChanged += args => {
    Console.WriteLine($"{args.employee.Name} changed from {args.oldManager.Name} to {args.newManager.Name}.");
};
```

**Reply by najak3d** [reply]
@HaloFour  - there's no viable way to fix that in "tooling" that fits.   We don't want it to show "full qualified Delegate name + parameters" -- that's TMI on the screen, because the delegate name can be fairly long.    But with "Action" - it's short, leaving plenty of room for parameter names.

I've been using Event examples, but better ones would be the many methods that require you to send it a delegate, where the Intellisense drama is more extreme, because the code doesn't auto-generate for you the footprint! ... you have to literally F12 to the delegate definition, cut/paste it, then come back to your code and create the method, looking at the pasted version.    If instead that delegate were an anonymous delegate (Action<...>), then you'd already see the parameters needed.

I'm going to stop barking up this tree.   I sense the resistance, and I'm sure my misconceptions about the value-of-this-proposal are skewed.   I'm a noob.  Y'all are the experts.

Thanks again for your dedication to protecting and enhancing the worlds-best-ever language - C#.

**Reply by najak3d** [reply]
I may start using the Tuple-notation to bridge-the-gap.   There is an advantage there in that you FORCE the consumer to use your named variables, and prevents confusion.   While for this Action<....> proposal, we wouldn't be ENFORCING same-named variables (as that could be mayhem)....   Which means that the author of the Action<..> could change the MEANING of parameters, without the consumers knowing it!

I think this is the biggest downside of our proposal here.  Is that in order for it to be SAFE, it would also have to FORCE same-names onto the consumers of the anonymous delegate to ensure their code is not misunderstanding the meaning of each argument.

While Tuples, although introducing an element of inconvenience (i.e. having to have an arbitrary variable named 'args' or 'data') -- it does help GUARANTEE safety by BREAKING the code of consumers, whenever a variable name is changed (as you WANT it to break them, to get their attention).   (This assumes they don't use the Item1/2/3 notation.)

So -- IMO -- THIS is the biggest argument against this proposal for delegates having named parameters -- it would ENCOURAGE MORE USE OF IT -- when it's not really a good idea to do so.

**Reply by najak3d** [reply]
So in that light -- I will GLADLY start using Tuple-notation, without looking back.   Tuple-notation is superior, and already implemented.

**Reply by najak3d** [reply]
OOOPS -- Just tested it... and Tuples suffers the same issue!   I'm new to those so didn't realize it.   The author of an event CAN change the parameter names, without it BREAKING consumer code -- leaving us with the same conundrum.

So both methods have the same conundrum, and in this manner, both have this safety-deficiency compared to using the "strong-named delegate" mechanism.

**Reply by najak3d** [reply]
Here is some code that demonstrates the weakness in using Tuples in this fashion:
(I have swapped the ORDER of names!)

```c#
  public event Action<(string oldManager, string newManager, string employee)> ManagerChanged;
  private void FalonSceneProcessor_ManagerChanged((string employee, string oldManager, string newManager) args)
  {
	  args.employee = "This Is Really 'oldManager'!!!!";
  }
```
So unless the Handler-method itself is also anonymous -- then this safety-issue remains.


**Is there a way to tell Visual-Studio to turn on a Warning (or treat as an error) the above code?**

**Reply by HaloFour** [reply]
Tuples were designed to work like parameter lists where the names are informational only.  If you want type safety, declare a type.

**Reply by najak3d** [reply]
The following bad code, issues No Warnings for me in VS 2022 (Preview):

![image](https://user-images.githubusercontent.com/10914515/155994269-b5d1c7b3-9cec-40dc-840b-c5b85fbaba67.png)

It would be nice if the "+=" assignment issued a warning for mismatched parameter names....  But, I see that it NEVER does this for other parameter name mismatches either -- so doesn't make sense for this to do it here either.

**EDIT:  In the case of Tuples, vs. other method signatures, there is better reason to do actual Parameter-Named Comparisons between the Tuple on the event signature, vs. the Tuple on the Event Handler method.  Tuple-Name Mismatches SHOULD be a warning....  as it's to be treated the SAME as if it were a full class definition (where property names are defined).**

**Reply by najak3d** [reply]
I'm stating this as some big "safety risk" - but it's not much different than that of method parameter lists (as you said), where an author of a method could easily just swap two like-typed parameters, which dangerously "maintains the parameter list signature" (same # parameters, same types -- but meanings have been changed or swapped).

And C# now provides a way to make this "type-safe" via the "{param name}: value" notation that you can use for method calling, if this type of safety is a concern.

Example:
```c#
public void OnManagerChanged(string oldMgr, string newMgr) { ... }

// changes to - and callers of this method have no idea you swapped parameter order
public void OnManagerChanged(string newMgr, string oldMgr) { ... }

// RESOLVED now in C# with calling code like this:
{
  OnManagerChanged(oldMgr: "oldGuy", newMgr: "newguy");
}
```

Both Tuples and 'named delegate types' alike still have this "parameter meaning swapped" danger.  You can make the same mistake with both alike.

The BENEFIT of Tuples vs 'named delegate types' is that at least the Intellisense tells you the issue without needing to F12 to definition.

**Reply by CyrusNajmabadi** [reply]
> there's no viable way to fix that in "tooling" that fits. We don't want it to show "full qualified Delegate name + parameters" -- that's TMI on the screen, because the delegate name can be fairly long. But with "Action" - it's short, leaving plenty of room for parameter names.

I believe there's a pr already to it to help with this scenario. It's not a compelling reason to change the language.

**Reply by CyrusNajmabadi** [reply]
> I think this is the biggest downside of our proposal here. Is that in order for it to be SAFE, it would also have to FORCE same-names onto the consumers of the anonymous delegate to ensure their code is not misunderstanding the meaning of each argument.

This is not how the language works for most scenarios. For example, most of the places you are talking about names today are not stringent about caller and callee matching. There are many times they do not have to match, and many places where you cannot force a break to happen when they do not match.

**Reply by najak3d** [reply]
I created a new discussion regarding making it a warning or error to match up two Tuples that have different parameter names.  I'd like to hear your thoughts on this.

https://github.com/dotnet/csharplang/discussions/5840
